Blocks: Avoid setAttributes on end-of-paragraph split

aduth opened the pull request:
This pull request is part of a performance audit of the paragraph block. It seeks to resolve an issue where we needlessly call `setAttributes` on a paragraph block when splitting into a new paragraph, and avoids potentially heavy processing within `RichText` to determine `before` and `after` fragments when it can be inferred by the position of the selection at the end/beginning of the `RichText` field whether the existing value can be used verbatim; thus also giving opportunity for the paragraph block to do a strict comparison against its own `content` attribute to determine whether `setAttributes` needs to be called.

This is one of two `setAttributes` calls which are currently invoked on the original paragraph when pressing enter at the end of its text. The other is explained at https://github.com/WordPress/gutenberg/pull/4956#discussion_r197473908 . 

**Testing instructions:**

Verify there are no regressions in the behavior of paragraph splitting.

Optional: Validate that `UPDATE_BLOCK_ATTRIBUTES` is dispatched only once ([erroneously](https://github.com/WordPress/gutenberg/pull/4956#discussion_r197473908), vs. twice on master) when pressing enter at the end of a paragraph block using [Redux DevTools Extension](https://github.com/zalmoxisus/redux-devtools-extension).

aduth committed:
Blocks: Avoid setAttributes on end-of-paragraph split

youknowriad reviewed (approved):
LGTM 👍

aduth commented:
I've a few things I'm looking into before merging this:

- Ensuring that there are no problems with calling `onSplit` directly rather than `restoreContentAndSplit`. My current conclusion is that we don't need `restoreContentAndSplit` at all, and it's proposed to be removed in #7618.
- Ensuring that there's no issue in not reaching the part of `splitContent` which considers `context.paste`. I'm still not entirely clear what it's responsible for doing, and am continue to explore further.

mtias commented on core-blocks/paragraph/index.js:
```diff
@@ -137,11 +139,53 @@ class ParagraphBlock extends Component {
+		} = this.props;
+
+		if ( after ) {
+			// After content should be appended to the set of spread blocks as
```
This comment was a bit hard to understand.

aduth replied:
>This comment was a bit hard to understand.

Updated in 2ea9199.

mtias reviewed (approved):
Looks good to me. Thanks for adding comments!

aduth committed:
Blocks: Preserve respect of insertBlocksAfter undefined

For locked blocks.

See also: https://github.com/WordPress/gutenberg/issues/6587

aduth committed:
Blocks: Clarify after split insertion coment

aduth committed:
Rich Text: Test edge as empty after / before fragment

Accommodates case where childNodes exist, but none which would qualify after filtering. Also preserves existing logic for detecting / handling paste context.

aduth commented:
Updates:

- Rebased to resolve conflicts.
- Restored respect of the `insertBlocksAfter` potentially being `undefined` for a locked template. See also #6587.
- Refactored `RichText` split once more: This time simplifying to handle edge detection at once, with a single path toward `restoreContentAndSplit` with handling of paste context. See extended commit message of 7a8ec43.

jorgefilipecosta commented:
> This pull request is part of a performance audit of the paragraph block. It seeks to resolve an issue where we needlessly call setAttributes on a paragraph block when splitting into a new paragraph.

I think we are stilling making a non required setAttributes call.

I added this paragraph as the content of the post:
```
<!-- wp:paragraph -->
<p>123456789</p>
<!-- /wp:paragraph -->
```
I opened redux dev tools in core/editor store.
I put the cursor between 5 and 6 and pressed enter.
I checked that we dispatched UPDATE_BLOCK_ATTRIBUTES with content equal to "123456789" (not needed). Then we inserted a block and then we updated the content of the original block.

It is possible to check that we are doing a non required UPDATE_BLOCK_ATTRIBUTES also by inserting many paragraphs without any content. To undo each paragraph creation we need to use undo two times.

aduth commented:
@jorgefilipecosta Nice catch. I can reproduce this. However, I think it's a totally separate issue. Namely, we're causing `UPDATE_BLOCK_ATTRIBUTES` to be called because even though we're explicitly handling the split on enter press, the event binding for `input` will cause `onChange` to be called with the editor content. We should probably prevent this for the enter key (or when otherwise knowing that there is other explicit handling for the input).
